### PR TITLE
Fix templated phpdoc return type

### PR DESCRIPTION
**Pull request by mvorisek**
When split between `@param` and `@psalm-param`, only the `@psalm-param` is used by psalm/phpstan, ie `class-string<T>` is used instead of (too wide) `class|class-string<T>`.

In the 2nd commit, I fixed all other simillar problems in the whole repo. I used `@.*(string.*\|class-s|class-string.*\|.*string)` regex to find the lines to check/fix.

### CODE REVIEW

**Comment by mvorisek** [comment]
I regenerated the psalm baseline locally, but the CI is still failing. Please regenerate it before merge.

**Comment by derrabus** [comment]
I don't believe that this change is correct. Can you elaborate what issue you're trying to fix here?

**Comment by mvorisek** [comment]
phpstan cannot otherwise infer the resulting type for `EntityManagerInterface::getClassMetadata()` method, see the no longer needed baseline ignore [1]

@derrabus why do you think this change is not correct?

see other methods phpdocs in `EntityManagerInterface` class, they are all, except this one, in the form I fixed this one to

[1] and in my usecase, I needed to add `Unable to resolve the template type T in call to method static method Doctrine\\Common\\Util\\ClassUtils::getClass\(\)` phpstan ignore /wo this PR

**Comment by derrabus** [comment]
> see the no longer needed baseline ignore
> […]
> why do you think this change is not correct?

See all the new Psalm errors that pop up after your change. 🙂

**Comment by mvorisek** [comment]
> > see the no longer needed baseline ignore
> > […]
> > why do you think this change is not correct?
> 
> See all the new Psalm errors that pop up after your change. 🙂

@derrabus looking into the psalm errors -> code: https://github.com/doctrine/orm/blob/2.15.4/lib/Doctrine/ORM/Cache.php#L42 it seems the are many places that accept a class name but are annotated as `string` instead of `class-string`.

I am not paid to maintain this project so I would be happy if this PR can be accepted and the psalm baseline regenerated (I tried myself on Windows, but the updated baseline did not fix all errors).

**Comment by derrabus** [comment]
> I am not paid to maintain this project

Neither am I. 🤷🏻‍♂️ 

> I would be happy if this PR can be accepted

I cannot merge a PR with a red CI and I won't fix other people's PRs in my freetime.

**Comment by SenseException** [comment]
@mvorisek What kind of problems do you have that you want to fix with this PR? Your last comment doesn't sound like that this is relevant to your daily work with ORM but more of an aid. We appreciate the effort people put into PRs as users (hopefully) appreciate the time we spent into the Doctrine project repositories on a regular basis.

We would like to know how this should be handled, because, as @derrabus already mentioned, this would rather be a hinderance to our also unpaid freetime if the failed build doesn't get fixed. I'd like to ask you to go the second half of what you've started.

File: lib/Doctrine/ORM/EntityManagerInterface.php
Comment (greg0ire) [inline lib/Doctrine/ORM/EntityManagerInterface.php:339]: `EntityManager::getClassMetadata()`  also supports aliased names and needs to do so until Persistence 2 support is removed:

https://github.com/doctrine/orm/blob/f7e4b61459692f9b747f40696e6bf72080390f2d/lib/Doctrine/ORM/EntityManager.php#L305-L316

Should the same be expected of any `EntityManagerInterface` implementation? If yes I'd say you're not so much fixing stuff as you are breaking it.

> Fix templated phpdoc return type 

Did you mean to write "parameter type" here?

**Review by greg0ire — CHANGES_REQUESTED** [review]
All 4 of your commit messages are unhelpful, or sometimes even plain wrong, it seems. As I probably have already told you in the past, you should read the contribution guidelines to figure out what to write, they're fairly extensive on the topic.

As for the the CI build, if you want to contribute to Doctrine, you should setup a proper development environment.

**Comment by mvorisek** [comment]
> What kind of problems do you have that you want to fix with this PR?

here is a phpstan repro: https://phpstan.org/r/b2cbe847-a748-4094-bc49-44bd3e200e12 explaining the need

notice the dumped type change when l6 is changed to `@psalm-param class-string<T> $className`

File: lib/Doctrine/ORM/EntityManagerInterface.php
Comment (mvorisek) [reply lib/Doctrine/ORM/EntityManagerInterface.php:339]: @greg0ire see https://github.com/doctrine/orm/pull/10849#issuecomment-1648223127 for repro, I meant "return type", as the "return type" is currently not correctly narrowed as the  "parameter type" does allow any `string`

**Comment by derrabus** [comment]
This function accepts strings that are not FQCN, so changing the parameter type to `class-string` would be wrong. But if we a `class-string<SomeClass>` is passed, the return type to expect is `ClassMetadata<SomeClass>`.

Can we solve this with a conditional return type?

File: lib/Doctrine/ORM/EntityManagerInterface.php
Comment (greg0ire) [reply lib/Doctrine/ORM/EntityManagerInterface.php:339]: Related: https://github.com/doctrine/orm/pull/10853

Comment (VincentLanglet) [reply lib/Doctrine/ORM/EntityManagerInterface.php:339]:
```
* @param string $className
     * @psalm-param class-string<T> $className
```

Might be a good way to deprecate string param for static analysis user
and solve https://github.com/phpstan/phpstan/issues/5175

**Comment by VincentLanglet** [comment]
> This function accepts strings that are not FQCN, so changing the parameter type to `class-string` would be wrong. But if we a `class-string<SomeClass>` is passed, the return type to expect is `ClassMetadata<SomeClass>`.
> 
> Can we solve this with a conditional return type?

There was an issue about this in phpstan
https://github.com/phpstan/phpstan/issues/5175

**Comment by mvorisek** [comment]
Closing as it seems this change is not wanted, at least for ORM 2.x.

**Comment by VincentLanglet** [comment]
> Closing as it seems this change is not wanted, at least for ORM 2.x.

I think the string support was deprecated so maybe the issue does not exist in 3.x anymore (or the phpdoc could be updated).

Need to check

**Pull request closed without merging**
